### PR TITLE
refactored all ops to remove old style ChargeForward methods

### PR DESCRIPTION
**Pull request by baykaner**

### CODE REVIEW

File: libs/ml/src/ops/abs.cpp [hunk 1]
```diff
@@ -111,12 +111,14 @@ std::vector<math::SizeType> Abs<TensorType>::ComputeOutputShape(
 }
 
 template <typename TensorType>
-OperationsCount Abs<TensorType>::ChargeForward() const
+std::pair<OperationsCount, math::SizeVector> Abs<TensorType>::ChargeForward(
+    std::vector<math::SizeVector> const &input_shapes)
 {
   assert(!this->batch_input_shapes_.empty());
```
Comment (evsmithx) [inline libs/ml/src/ops/abs.cpp]: Now that we are no longer using batch_input_shapes_ in this method I think we can get rid of the asserts

File: libs/ml/src/ops/avg_pool_1d.cpp [hunk 2]
```diff
@@ -180,17 +180,20 @@ std::vector<math::SizeType> AvgPool1D<TensorType>::ComputeOutputShape(
...
                                          this->batch_output_shape_.at(2);
-  OperationsCount cost =
+  OperationsCount op_cnt =
       fetch::ml::charge_estimation::ops::DIVISION_PER_ELEMENT * num_output_shape_ops +
       fetch::ml::charge_estimation::ops::ADDITION_PER_ELEMENT * num_output_shape_ops *
           this->kernel_size_;
-  return cost;
+
+  auto output_shape = ComputeOutputShape(input_shapes);
+  return std::make_pair(op_cnt, output_shape);
```
Comment (evsmithx) [inline libs/ml/src/ops/avg_pool_1d.cpp:194]: ChargeForward should now return the cost for the full batch, not the per-datapoint cost. Assuming the old method returned the correct per-datapoint cost then this method should return op_cnt * batch_size

File: libs/ml/src/ops/avg_pool_2d.cpp [hunk 3]
```diff
@@ -199,17 +199,20 @@ std::vector<math::SizeType> AvgPool2D<TensorType>::ComputeOutputShape(
...
       this->batch_output_shape_.at(2) * this->batch_output_shape_.at(3);
-  auto cost = static_cast<OperationsCount>(
+  auto op_cnt = static_cast<OperationsCount>(
       fetch::ml::charge_estimation::ops::DIVISION_PER_ELEMENT * num_output_shape_ops +
       fetch::ml::charge_estimation::ops::ADDITION_PER_ELEMENT * num_output_shape_ops *
           static_cast<OperationsCount>(this->kernel_size_ * this->kernel_size_));
-  return cost;
+
+  auto output_shape = ComputeOutputShape(input_shapes);
+  return std::make_pair(op_cnt, output_shape);
```
Comment (evsmithx) [inline libs/ml/src/ops/avg_pool_2d.cpp:214]: op_cnt *= batch size again

File: libs/ml/src/ops/convolution_1d.cpp [hunk 4]
```diff
@@ -411,28 +411,30 @@ void Convolution1D<TensorType>::ReverseFillOutput(TensorType &gemm_output, Tenso
...
-  OperationsCount cost = horizontal_stride_width * horizontal_stride_height *
-                         vertical_stride_width *
-                         fetch::ml::charge_estimation::ops::MULTIPLICATION_PER_ELEMENT;
+  OperationsCount op_cnt = horizontal_stride_width * horizontal_stride_height *
+                           vertical_stride_width *
+                           fetch::ml::charge_estimation::ops::MULTIPLICATION_PER_ELEMENT;
 
-  return cost;
+  auto output_shape = ComputeOutputShape(input_shapes);
+  return std::make_pair(op_cnt, output_shape);
```
Comment (evsmithx) [inline libs/ml/src/ops/convolution_1d.cpp]: *= batch size again (I think)

File: libs/ml/src/ops/convolution_2d.cpp [hunk 5]
```diff
@@ -449,32 +449,32 @@ void Convolution2D<TensorType>::ReverseFillOutput(TensorType &gemm_output, Tenso
...
-  OperationsCount cost = horizontal_stride_width * horizontal_stride_height *
-                         vertical_stride_width *
-                         fetch::ml::charge_estimation::ops::MULTIPLICATION_PER_ELEMENT;
+  OperationsCount op_cnt = horizontal_stride_width * horizontal_stride_height *
+                           vertical_stride_width *
+                           fetch::ml::charge_estimation::ops::MULTIPLICATION_PER_ELEMENT;
 
-  return cost;
+  auto output_shape = ComputeOutputShape(input_shapes);
+  return std::make_pair(op_cnt, output_shape);
```
Comment (evsmithx) [inline libs/ml/src/ops/convolution_2d.cpp]: batch size

File: libs/ml/src/ops/reshape.cpp [hunk 6]
```diff
@@ -124,14 +124,17 @@ std::vector<math::SizeType> Reshape<TensorType>::ComputeOutputShape(
...
-  assert(!this->batch_output_shape_.empty());
-  OperationsCount cost = fetch::ml::charge_estimation::ops::RESHAPE_PER_ELEMENT *
-                             this->TotalElementsIn({this->batch_output_shape_}) +
-                         fetch::ml::charge_estimation::ops::ASSIGN_PER_ELEMENT *
-                             this->TotalElementsIn({this->batch_output_shape_});
-  return cost;
+  assert(!this->batch_input_shapes_.empty());
+
+  OperationsCount op_cnt = (fetch::ml::charge_estimation::ops::RESHAPE_PER_ELEMENT +
+                            fetch::ml::charge_estimation::ops::RESHAPE_PER_ELEMENT) *
```
Comment (evsmithx) [inline libs/ml/src/ops/reshape.cpp]: This should be ASSIGN_PER_ELEMENT?

**Review by evsmithx — CHANGES_REQUESTED** [review]

File: libs/ml/src/ops/abs.cpp (hunk 1 above)
Comment (baykaner) [reply libs/ml/src/ops/abs.cpp]: good point

File: libs/ml/src/ops/avg_pool_1d.cpp (hunk 2 above)
Comment (baykaner) [reply libs/ml/src/ops/avg_pool_1d.cpp:194]: input shapes contains that info though right? so if i just swap out the 'batch shapes' for input shapes it'll include the batch size for free i think.

File: libs/ml/src/ops/avg_pool_2d.cpp (hunk 3 above)
Comment (baykaner) [reply libs/ml/src/ops/avg_pool_2d.cpp:214]: as above

File: libs/ml/src/ops/convolution_1d.cpp (hunk 4 above)
Comment (baykaner) [reply libs/ml/src/ops/convolution_1d.cpp]: here I've already done the swap

File: libs/ml/src/ops/convolution_2d.cpp (hunk 5 above)
Comment (baykaner) [reply libs/ml/src/ops/convolution_2d.cpp]: also here

File: libs/ml/src/ops/reshape.cpp (hunk 6 above)
Comment (baykaner) [reply libs/ml/src/ops/reshape.cpp]: I haven't fine-tuned it. So either is fine at this stage.

File: libs/ml/src/ops/avg_pool_1d.cpp (hunk 2 above)
Comment (evsmithx) [reply libs/ml/src/ops/avg_pool_1d.cpp:194]: Yep, but you're using batch_output_shape_ above!

File: libs/ml/src/ops/avg_pool_2d.cpp (hunk 3 above)
Comment (evsmithx) [reply libs/ml/src/ops/avg_pool_2d.cpp:214]: Again, you're using batch_output_shape!

File: libs/ml/src/ops/convolution_1d.cpp (hunk 4 above)
Comment (evsmithx) [reply libs/ml/src/ops/convolution_1d.cpp]: Yep, agreed

File: libs/ml/src/ops/reshape.cpp (hunk 6 above)
Comment (evsmithx) [reply libs/ml/src/ops/reshape.cpp]: Okay, I just noticed it had changed from the previous implementation!

File: libs/ml/src/ops/avg_pool_2d.cpp (hunk 3 above)
Comment (baykaner) [reply libs/ml/src/ops/avg_pool_2d.cpp:214]: not anymore - i pushed changes

File: libs/ml/src/ops/avg_pool_1d.cpp (hunk 2 above)
Comment (evsmithx) [reply libs/ml/src/ops/avg_pool_1d.cpp:194]: Github fail, it's fixed

**Review by evsmithx — DISMISSED** [review]

**Review by evsmithx — DISMISSED** [review]

**Review by MissingNO57 — APPROVED** [review]
LGTM

**Pull request merged**